{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "import time\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "\n",
    "\n",
    "from helpers.consensus_utils import select_best_alignment\n",
    "from helpers.fastq import concatenate_fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pipeline_consensus(input_name: str, preprocess: bool = False, windows: bool = False): \n",
    "    input_fastq_path = os.path.join(\"assets\", \"input\", input_name)\n",
    "\n",
    "    # In case the fastq are not yet concatenated, we do so here before proceeding with the rest of the code\n",
    "    if preprocess : \n",
    "        if not os.path.isdir(input_fastq_path) : \n",
    "            raise NotADirectoryError\n",
    "        \n",
    "        concatenate_fastq(input_fastq_path, input_fastq_path + \".fastq\")\n",
    "\n",
    "    input_fastq_path = input_fastq_path + \".fastq\"\n",
    "    if not os.path.isfile(input_fastq_path) : \n",
    "        raise FileNotFoundError\n",
    "    \n",
    "    output_dir = os.path.join(\"assets\", \"output\", input_name)\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "    # Split the file into top 20% and remaining 80%\n",
    "    top_sequences_path, remaining_sequences_path = split_fastq(input_fastq_path, output_dir, input_name, percentile=20)\n",
    "\n",
    "    total_time_taken = 0\n",
    "\n",
    "    # Step 1: Align and generate consensus from top 20% sequences\n",
    "    top_paf_path = os.path.join(output_dir, f\"{input_name}_top20_reads.paf\")\n",
    "    top_consensus_path = os.path.join(output_dir, f\"{input_name}_top20_consensus.fasta\")\n",
    "    minimap2_command = f\"minimap2 -x ava-ont {top_sequences_path} {top_sequences_path} > {top_paf_path}\"\n",
    "    if windows : \n",
    "        minimap2_command = minimap2_command.replace('\\\\','/')\n",
    "    print(\"Running read alignment with minimap2 on top 20% sequences...\")\n",
    "    _, minimap2_time = run_command(minimap2_command, windows)\n",
    "    total_time_taken += minimap2_time\n",
    "\n",
    "    racon_command = f\"racon -m 8 -x -6 -g -8 -w 500 {top_sequences_path} {top_paf_path} {top_sequences_path} > {top_consensus_path}\"\n",
    "    if windows : \n",
    "        racon_command = racon_command.replace('\\\\','/')\n",
    "    print(\"Generating consensus sequence with racon on top 20% sequences...\")\n",
    "    _, racon_time = run_command(racon_command, windows)\n",
    "    total_time_taken += racon_time\n",
    "\n",
    "    # Step 1 bis: If the outputed racon file contains more than one sequence, we take the longest one\n",
    "    consensus_sequences = list(SeqIO.parse(top_consensus_path, \"fasta\"))\n",
    "    if len(consensus_sequences) > 1:\n",
    "        print(f\"Multiple sequences found in {top_consensus_path}. Selecting the best alignment...\")\n",
    "        best_sequence = select_best_alignment(consensus_sequences, consensus_sequences[0])  # Choose the first sequence as the reference\n",
    "        SeqIO.write(best_sequence, top_consensus_path, \"fasta\")\n",
    "\n",
    "    # Step 2: Align the remaining 80% sequences to the top 20% consensus\n",
    "    remaining_paf_path = os.path.join(output_dir, f\"{input_name}_remaining80_reads.paf\")\n",
    "    final_consensus_path = os.path.join(output_dir, f\"{input_name}_final_consensus.fasta\")\n",
    "    minimap2_command = f\"minimap2 -x map-ont {top_consensus_path} {remaining_sequences_path} > {remaining_paf_path}\"\n",
    "    if windows : \n",
    "        minimap2_command = minimap2_command.replace('\\\\', '/')\n",
    "    print(\"Running read alignment with minimap2 on remaining 80% sequences...\")\n",
    "    _, minimap2_time = run_command(minimap2_command, windows)\n",
    "    total_time_taken += minimap2_time\n",
    "\n",
    "    # Step 3: Generate the final consensus sequence with racon\n",
    "    racon_command = f\"racon -m 8 -x -6 -g -8 -w 500 {remaining_sequences_path} {remaining_paf_path} {top_consensus_path} > {final_consensus_path}\"\n",
    "    if windows : \n",
    "        racon_command = racon_command.replace('\\\\','/')\n",
    "    print(\"Generating final consensus sequence with racon...\")\n",
    "    _, racon_time = run_command(racon_command, windows)\n",
    "    total_time_taken += racon_time\n",
    "\n",
    "    # Print out the total time for each step\n",
    "    print(f\"Minimap2 alignment took {minimap2_time:.2f} seconds.\")\n",
    "    print(f\"Total Racon iterations took {total_time_taken - minimap2_time:.2f} seconds.\")\n",
    "\n",
    "    # Print out the total time for the pipeline\n",
    "    print(f\"Total time taken for the pipeline: {total_time_taken:.2f} seconds.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pipeline_identification(input_name: str, database: str = None, windows: bool = False) :\n",
    "    input_fasta_path = os.path.join(\"assets\", \"output\", input_name, input_name + \"_final_consensus.fasta\")\n",
    "    output_blastn = os.path.join(\"assets\", \"output\", \"blastn\", input_name)\n",
    "    os.makedirs(output_blastn, exist_ok=True)\n",
    "\n",
    "    if database == None : \n",
    "        for db in {\"ITS_seqs\", \"matK_seqs\", \"psbA-trnH_seqs\", \"rbcL_seqs\"} :\n",
    "            output_blastn_path = os.path.join(output_blastn, db + \".txt\")\n",
    "            blastn_cmd = f\"blastn -query {input_fasta_path} -db {db} -out {output_blastn_path} -max_target_seqs 5\"\n",
    "            if windows : \n",
    "                blastn_cmd = blastn_cmd.replace(\"\\\\\", \"/\")\n",
    "            run_command(blastn_cmd)\n",
    "    else :\n",
    "        output_blastn_path = os.path.join(output_blastn, database + \".txt\")\n",
    "        blastn_cmd = f\"blastn -query {input_fasta_path} -db {database} -out {output_blastn_path} -max_target_seqs 5\"\n",
    "        if windows : \n",
    "                blastn_cmd = blastn_cmd.replace(\"\\\\\", \"/\")\n",
    "        run_command(blastn_cmd)\n",
    "\n",
    "    print(\"You can find identification ouput at \", output_blastn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(input_name: str, preprocess: bool = False, database: str = None, windows: bool = False) :\n",
    "    print(\"Debuting consensus sequencing : \") \n",
    "    pipeline_consensus(input_name, preprocess, windows)\n",
    "    print(\"Consensus sequencing finished\")\n",
    "    print(\"Debuting Identification with blastn : \")\n",
    "    pipeline_identification(input_name, database, windows)\n",
    "    print(\"Identification finished\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "main(\"rbcL_Qiagen_tomato_5000\", False, \"rbcL_seqs\", True)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
