{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "import time\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "\n",
    "\n",
    "from helpers.consensus_utils import select_best_alignment\n",
    "from helpers.fastq import concatenate_fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_command(command: str, windows: bool = False) -> (subprocess.CompletedProcess, float):\n",
    "    if windows : \n",
    "        command = \"wsl \" + command\n",
    "    print(f\"Running command: {command}\")\n",
    "\n",
    "    start_time = time.time()\n",
    "    result = subprocess.run(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)\n",
    "    elapsed_time = time.time() - start_time\n",
    "\n",
    "    print(f\"Command '{command}' took {elapsed_time:.2f} seconds to run.\")\n",
    "    if result.stdout:\n",
    "        print(result.stdout)\n",
    "    if result.stderr:\n",
    "        print(\"Error:\", result.stderr)\n",
    "    return result, elapsed_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_fastq(input_path, output_dir, base_name, percentile=20):\n",
    "    # Read sequences and sort by length\n",
    "    sequences = list(SeqIO.parse(input_path, \"fastq\"))\n",
    "    sequences.sort(key=lambda x: len(x), reverse=True)\n",
    "\n",
    "    # Split into top percentile and remaining sequences\n",
    "    split_index = len(sequences) * percentile // 100\n",
    "    top_sequences = sequences[:split_index]\n",
    "    remaining_sequences = sequences[split_index:]\n",
    "\n",
    "    # Write to separate files\n",
    "    top_sequences_path = os.path.join(output_dir, f\"{base_name}_top{percentile}.fastq\")\n",
    "    remaining_sequences_path = os.path.join(output_dir, f\"{base_name}_remaining{100-percentile}.fastq\")\n",
    "    SeqIO.write(top_sequences, top_sequences_path, \"fastq\")\n",
    "    SeqIO.write(remaining_sequences, remaining_sequences_path, \"fastq\")\n",
    "\n",
    "    return top_sequences_path, remaining_sequences_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pipeline_consensus(input_name: str, preprocess: bool = False, windows: bool = False): \n",
    "    input_fastq_path = os.path.join(\"assets\", \"input\", input_name)\n",
    "\n",
    "    # In case the fastq are not yet concatenated, we do so here before proceeding with the rest of the code\n",
    "    if preprocess : \n",
    "        if not os.path.isdir(input_fastq_path) : \n",
    "            raise NotADirectoryError\n",
    "        \n",
    "        concatenate_fastq(input_fastq_path, input_fastq_path + \".fastq\")\n",
    "\n",
    "    input_fastq_path = input_fastq_path + \".fastq\"\n",
    "    if not os.path.isfile(input_fastq_path) : \n",
    "        raise FileNotFoundError\n",
    "    \n",
    "    output_dir = os.path.join(\"assets\", \"output\", input_name)\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "    # Split the file into top 20% and remaining 80%\n",
    "    top_sequences_path, remaining_sequences_path = split_fastq(input_fastq_path, output_dir, input_name, percentile=20)\n",
    "\n",
    "    total_time_taken = 0\n",
    "\n",
    "    # Step 1: Align and generate consensus from top 20% sequences\n",
    "    top_paf_path = os.path.join(output_dir, f\"{input_name}_top20_reads.paf\")\n",
    "    top_consensus_path = os.path.join(output_dir, f\"{input_name}_top20_consensus.fasta\")\n",
    "    minimap2_command = f\"minimap2 -x ava-ont {top_sequences_path} {top_sequences_path} > {top_paf_path}\"\n",
    "    if windows : \n",
    "        minimap2_command = minimap2_command.replace('\\\\','/')\n",
    "    print(\"Running read alignment with minimap2 on top 20% sequences...\")\n",
    "    _, minimap2_time = run_command(minimap2_command, windows)\n",
    "    total_time_taken += minimap2_time\n",
    "\n",
    "    racon_command = f\"racon -m 8 -x -6 -g -8 -w 500 {top_sequences_path} {top_paf_path} {top_sequences_path} > {top_consensus_path}\"\n",
    "    if windows : \n",
    "        racon_command = racon_command.replace('\\\\','/')\n",
    "    print(\"Generating consensus sequence with racon on top 20% sequences...\")\n",
    "    _, racon_time = run_command(racon_command, windows)\n",
    "    total_time_taken += racon_time\n",
    "\n",
    "    # Step 1 bis: If the outputed racon file contains more than one sequence, we take the longest one\n",
    "    consensus_sequences = list(SeqIO.parse(top_consensus_path, \"fasta\"))\n",
    "    if len(consensus_sequences) > 1:\n",
    "        print(f\"Multiple sequences found in {top_consensus_path}. Selecting the best alignment...\")\n",
    "        best_sequence = select_best_alignment(consensus_sequences, consensus_sequences[0])  # Choose the first sequence as the reference\n",
    "        SeqIO.write(best_sequence, top_consensus_path, \"fasta\")\n",
    "\n",
    "    # Step 2: Align the remaining 80% sequences to the top 20% consensus\n",
    "    remaining_paf_path = os.path.join(output_dir, f\"{input_name}_remaining80_reads.paf\")\n",
    "    final_consensus_path = os.path.join(output_dir, f\"{input_name}_final_consensus.fasta\")\n",
    "    minimap2_command = f\"minimap2 -x map-ont {top_consensus_path} {remaining_sequences_path} > {remaining_paf_path}\"\n",
    "    if windows : \n",
    "        minimap2_command = minimap2_command.replace('\\\\', '/')\n",
    "    print(\"Running read alignment with minimap2 on remaining 80% sequences...\")\n",
    "    _, minimap2_time = run_command(minimap2_command, windows)\n",
    "    total_time_taken += minimap2_time\n",
    "\n",
    "    # Step 3: Generate the final consensus sequence with racon\n",
    "    racon_command = f\"racon -m 8 -x -6 -g -8 -w 500 {remaining_sequences_path} {remaining_paf_path} {top_consensus_path} > {final_consensus_path}\"\n",
    "    if windows : \n",
    "        racon_command = racon_command.replace('\\\\','/')\n",
    "    print(\"Generating final consensus sequence with racon...\")\n",
    "    _, racon_time = run_command(racon_command, windows)\n",
    "    total_time_taken += racon_time\n",
    "\n",
    "    # Print out the total time for each step\n",
    "    print(f\"Minimap2 alignment took {minimap2_time:.2f} seconds.\")\n",
    "    print(f\"Total Racon iterations took {total_time_taken - minimap2_time:.2f} seconds.\")\n",
    "\n",
    "    # Print out the total time for the pipeline\n",
    "    print(f\"Total time taken for the pipeline: {total_time_taken:.2f} seconds.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pipeline_identification(input_name: str, database: str = None, windows: bool = False) :\n",
    "    input_fasta_path = os.path.join(\"assets\", \"output\", input_name, input_name + \"_final_consensus.fasta\")\n",
    "    output_blastn = os.path.join(\"assets\", \"output\", \"blastn\", input_name)\n",
    "    os.makedirs(output_blastn, exist_ok=True)\n",
    "\n",
    "    if database == None : \n",
    "        for db in {\"ITS_seqs\", \"matK_seqs\", \"psbA-trnH_seqs\", \"rbcL_seqs\"} :\n",
    "            output_blastn_path = os.path.join(output_blastn, db + \".txt\")\n",
    "            blastn_cmd = f\"blastn -query {input_fasta_path} -db {db} -out {output_blastn_path} -max_target_seqs 5\"\n",
    "            if windows : \n",
    "                blastn_cmd = blastn_cmd.replace(\"\\\\\", \"/\")\n",
    "            run_command(blastn_cmd)\n",
    "    else :\n",
    "        output_blastn_path = os.path.join(output_blastn, database + \".txt\")\n",
    "        blastn_cmd = f\"blastn -query {input_fasta_path} -db {database} -out {output_blastn_path} -max_target_seqs 5\"\n",
    "        if windows : \n",
    "                blastn_cmd = blastn_cmd.replace(\"\\\\\", \"/\")\n",
    "        run_command(blastn_cmd)\n",
    "\n",
    "    print(\"You can find identification ouput at \", output_blastn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(input_name: str, preprocess: bool = False, database: str = None, windows: bool = False) :\n",
    "    print(\"Debuting consensus sequencing : \") \n",
    "    pipeline_consensus(input_name, preprocess, windows)\n",
    "    print(\"Consensus sequencing finished\")\n",
    "    print(\"Debuting Identification with blastn : \")\n",
    "    pipeline_identification(input_name, database, windows)\n",
    "    print(\"Identification finished\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Debuting consensus sequencing : \n",
      "Running read alignment with minimap2 on top 20% sequences...\n",
      "Running command: wsl minimap2 -x ava-ont assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_reads.paf\n",
      "Command 'wsl minimap2 -x ava-ont assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_reads.paf' took 2.38 seconds to run.\n",
      "Error: [M::mm_idx_gen::0.056*0.37] collected minimizers\n",
      "[M::mm_idx_gen::0.066*0.66] sorted minimizers\n",
      "[M::main::0.066*0.66] loaded/built the index for 1000 target sequence(s)\n",
      "[M::mm_mapopt_update::0.068*0.67] mid_occ = 608\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 5; is_hpc: 0; #seq: 1000\n",
      "[M::mm_idx_stat::0.069*0.68] distinct minimizers: 61942 (78.57% are singletons); average occurrences: 3.831; average spacing: 2.966; total length: 703802\n",
      "[M::worker_pipeline::2.213*2.60] mapped 1000 sequences\n",
      "[M::main] Version: 2.24-r1122\n",
      "[M::main] CMD: minimap2 -x ava-ont assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq\n",
      "[M::main] Real time: 2.217 sec; CPU: 5.762 sec; Peak RSS: 0.062 GB\n",
      "\n",
      "Generating consensus sequence with racon on top 20% sequences...\n",
      "Running command: wsl racon -m 8 -x -6 -g -8 -w 500 assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_reads.paf assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta\n",
      "Command 'wsl racon -m 8 -x -6 -g -8 -w 500 assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_reads.paf assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta' took 1.73 seconds to run.\n",
      "Error: [racon::Polisher::initialize] loaded target sequences 0.014578 s\n",
      "[racon::Polisher::initialize] loaded sequences 0.009260 s\n",
      "[racon::Polisher::initialize] loaded overlaps 0.867885 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=>                  ] 0.046746 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==>                 ] 0.050231 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===>                ] 0.053768 s\n",
      "[racon::Polisher::initialize] aligning overlaps [====>               ] 0.057358 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=====>              ] 0.060808 s\n",
      "[racon::Polisher::initialize] aligning overlaps [======>             ] 0.064302 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=======>            ] 0.067799 s\n",
      "[racon::Polisher::initialize] aligning overlaps [========>           ] 0.071250 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=========>          ] 0.075259 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==========>         ] 0.079432 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===========>        ] 0.083088 s\n",
      "[racon::Polisher::initialize] aligning overlaps [============>       ] 0.086705 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=============>      ] 0.090277 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==============>     ] 0.093890 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===============>    ] 0.097512 s\n",
      "[racon::Polisher::initialize] aligning overlaps [================>   ] 0.101319 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=================>  ] 0.104676 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==================> ] 0.108005 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===================>] 0.111326 s\n",
      "[racon::Polisher::initialize] aligning overlaps [====================] 0.115277 s\n",
      "[racon::Polisher::initialize] transformed data into windows 0.001215 s\n",
      "[racon::Polisher::polish] generating consensus [=>                  ] 0.359632 s\n",
      "[racon::Polisher::polish] generating consensus [==>                 ] 0.514917 s\n",
      "[racon::Polisher::polish] generating consensus [===>                ] 0.556813 s\n",
      "[racon::Polisher::polish] generating consensus [====>               ] 0.561525 s\n",
      "[racon::Polisher::polish] generating consensus [=====>              ] 0.565886 s\n",
      "[racon::Polisher::polish] generating consensus [======>             ] 0.566913 s\n",
      "[racon::Polisher::polish] generating consensus [=======>            ] 0.569488 s\n",
      "[racon::Polisher::polish] generating consensus [========>           ] 0.569752 s\n",
      "[racon::Polisher::polish] generating consensus [=========>          ] 0.569772 s\n",
      "[racon::Polisher::polish] generating consensus [==========>         ] 0.570632 s\n",
      "[racon::Polisher::polish] generating consensus [===========>        ] 0.570679 s\n",
      "[racon::Polisher::polish] generating consensus [============>       ] 0.570745 s\n",
      "[racon::Polisher::polish] generating consensus [=============>      ] 0.571873 s\n",
      "[racon::Polisher::polish] generating consensus [==============>     ] 0.572096 s\n",
      "[racon::Polisher::polish] generating consensus [===============>    ] 0.572115 s\n",
      "[racon::Polisher::polish] generating consensus [================>   ] 0.572131 s\n",
      "[racon::Polisher::polish] generating consensus [=================>  ] 0.573396 s\n",
      "[racon::Polisher::polish] generating consensus [==================> ] 0.573493 s\n",
      "[racon::Polisher::polish] generating consensus [===================>] 0.573614 s\n",
      "[racon::Polisher::polish] generating consensus [====================] 0.574783 s\n",
      "[racon::Polisher::] total = 1.583685 s\n",
      "\n",
      "Multiple sequences found in assets\\output\\rbcL_Qiagen_tomato_5000\\rbcL_Qiagen_tomato_5000_top20_consensus.fasta. Selecting the best alignment...\n",
      "Running read alignment with minimap2 on remaining 80% sequences...\n",
      "Running command: wsl minimap2 -x map-ont assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80.fastq > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80_reads.paf\n",
      "Command 'wsl minimap2 -x map-ont assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80.fastq > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80_reads.paf' took 0.45 seconds to run.\n",
      "Error: [M::mm_idx_gen::0.022*0.48] collected minimizers\n",
      "[M::mm_idx_gen::0.024*0.59] sorted minimizers\n",
      "[M::main::0.024*0.59] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.024*0.59] mid_occ = 10\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.024*0.59] distinct minimizers: 129 (100.00% are singletons); average occurrences: 1.000; average spacing: 5.961; total length: 769\n",
      "[M::worker_pipeline::0.293*0.24] mapped 4000 sequences\n",
      "[M::main] Version: 2.24-r1122\n",
      "[M::main] CMD: minimap2 -x map-ont assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80.fastq\n",
      "[M::main] Real time: 0.296 sec; CPU: 0.070 sec; Peak RSS: 0.007 GB\n",
      "\n",
      "Generating final consensus sequence with racon...\n",
      "Running command: wsl racon -m 8 -x -6 -g -8 -w 500 assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80_reads.paf assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_final_consensus.fasta\n",
      "Command 'wsl racon -m 8 -x -6 -g -8 -w 500 assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80_reads.paf assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_final_consensus.fasta' took 3.63 seconds to run.\n",
      "Error: [racon::Polisher::initialize] loaded target sequences 0.000546 s\n",
      "[racon::Polisher::initialize] loaded sequences 0.028508 s\n",
      "[racon::Polisher::initialize] loaded overlaps 0.006448 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=>                  ] 0.020370 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==>                 ] 0.027935 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===>                ] 0.035427 s\n",
      "[racon::Polisher::initialize] aligning overlaps [====>               ] 0.042287 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=====>              ] 0.048720 s\n",
      "[racon::Polisher::initialize] aligning overlaps [======>             ] 0.054460 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=======>            ] 0.060561 s\n",
      "[racon::Polisher::initialize] aligning overlaps [========>           ] 0.065586 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=========>          ] 0.070574 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==========>         ] 0.075239 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===========>        ] 0.078862 s\n",
      "[racon::Polisher::initialize] aligning overlaps [============>       ] 0.083370 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=============>      ] 0.087579 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==============>     ] 0.092845 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===============>    ] 0.097717 s\n",
      "[racon::Polisher::initialize] aligning overlaps [================>   ] 0.101200 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=================>  ] 0.104629 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==================> ] 0.107792 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===================>] 0.110278 s\n",
      "[racon::Polisher::initialize] aligning overlaps [====================] 0.111659 s\n",
      "[racon::Polisher::initialize] transformed data into windows 0.001207 s\n",
      "[racon::Polisher::polish] generated consensus 3.340308 s\n",
      "[racon::Polisher::] total = 3.490351 s\n",
      "\n",
      "Minimap2 alignment took 0.45 seconds.\n",
      "Total Racon iterations took 7.74 seconds.\n",
      "Total time taken for the pipeline: 8.19 seconds.\n",
      "Consensus sequencing finished\n",
      "Debuting Identification with blastn\n",
      "Running command: blastn -query assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_final_consensus.fasta -db rbcL_seqs -out assets/output/blastn/rbcL_Qiagen_tomato_5000/rbcL_seqs.txt -max_target_seqs 5\n",
      "Command 'blastn -query assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_final_consensus.fasta -db rbcL_seqs -out assets/output/blastn/rbcL_Qiagen_tomato_5000/rbcL_seqs.txt -max_target_seqs 5' took 1.14 seconds to run.\n",
      "You can find identification ouput at  assets\\output\\blastn\\rbcL_Qiagen_tomato_5000\n",
      "Identification finished\n"
     ]
    }
   ],
   "source": [
    "main(\"rbcL_Qiagen_tomato_5000\", False, \"rbcL_seqs\", True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
